AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Task/AWSAuthConfirmSignInTask.swift

```diff
@@ -0,0 +1,115 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+import Foundation
+import Amplify
+import AWSPluginsCore
+
+class AWSAuthConfirmSignInTask: AuthConfirmSignInTask {
+
+    private let request: AuthConfirmSignInRequest
+    private let authStateMachine: AuthStateMachine
+    private var stateListenerToken: AuthStateMachineToken?
+    
+    init(_ request: AuthConfirmSignInRequest, stateMachine: AuthStateMachine) {
+        self.request = request
+        self.authStateMachine = stateMachine
+        super.init(eventName: HubPayload.EventName.Auth.confirmSignInAPI)
+    }
+    
+    override var value: AuthSignInResult {
+        get async throws {
+            return try await execute()
+        }
+    }
+
+    private func execute() async throws -> AuthSignInResult {
+        if let validationError = request.hasError() {
+            dispatch(result: .failure(validationError))
+            throw validationError
+        }
+        
+        return try await withCheckedThrowingContinuation{ (continuation: CheckedContinuation<AuthSignInResult, Error>) in
+            let invalidStateError = AuthError.invalidState("User is not attempting signIn operation",
+                                                           AuthPluginErrorConstants.invalidStateError, nil)
+
+            stateListenerToken = authStateMachine.listen { [weak self] state in
+                guard let self = self, case .configured(let authNState, let authZState) = state else {
+                    self?.dispatch(result: .failure(invalidStateError))
+                    self?.cancelToken()
+                    continuation.resume(throwing: invalidStateError)
+                    return
+                }
+                switch authNState {
+                case .signedIn:
+                    if case .sessionEstablished = authZState {
+                        let result = AuthSignInResult(nextStep: .done)
+                        self.dispatch(result: .success(result))
+                        self.cancelToken()
+                        continuation.resume(returning: result)
+                    }
+                case .error(let error):
+                    let authError = AuthError.unknown("Sign in reached an error state", error)
+                    self.dispatch(result: .failure(authError))
+                    self.cancelToken()
+                    continuation.resume(throwing: authError)
+                case .signingIn(let signInState):
```

Comment:
Does this work if we retry after a failed confirmation?

Reply:
I added additional unit tests for to handle second sign in confirmation after failure and second signIn after first failure.  Both unit tests passed and succeeded in manual testing.  The auth integration tests will need to be updated.  I will update the integration tests once other dependent operations are also migrated over in a separate PR.